extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java

```diff
@@ -129,46 +152,63 @@ private static QuarkusSecurityIdentity augmentIdentity(SecurityIdentity security
         return cf;
     }
 
-    private void processSuccessfulAuthentication(RoutingContext context, CompletableFuture<SecurityIdentity> cf,
+    private void processSuccessfulAuthentication(RoutingContext context,
+            CompletableFuture<SecurityIdentity> cf,
             AccessToken result, SecurityIdentity securityIdentity) {
-        removeSessionCookie(context);
+        removeCookie(context, SESSION_COOKIE_NAME);
         CookieImpl cookie = new CookieImpl(SESSION_COOKIE_NAME, new StringBuilder(result.opaqueIdToken())
-                .append(SESSION_COOKIE_DELIM)
+                .append(COOKIE_DELIM)
                 .append(result.opaqueAccessToken())
-                .append(SESSION_COOKIE_DELIM)
+                .append(COOKIE_DELIM)
                 .append(result.opaqueRefreshToken()).toString());
 
         cookie.setMaxAge(result.idToken().getInteger("exp"));
         cookie.setSecure(context.request().isSSL());
         cookie.setHttpOnly(true);
-
         context.response().addCookie(cookie);
+
         cf.complete(augmentIdentity(securityIdentity, result.opaqueAccessToken(),
                 result.opaqueRefreshToken()));
     }
 
-    private String generateState(RoutingContext context) {
-        CookieImpl cookie = new CookieImpl(STATE_COOKIE_NAME, UUID.randomUUID().toString());
+    private String getDynamicPath(RoutingContext context, URI absoluteUri) {
+        if (config.authentication.redirectPath.isPresent()) {
+            String redirectPath = config.authentication.redirectPath.get();
+            String requestPath = absoluteUri.getRawPath();
+            if (requestPath.startsWith(redirectPath) && requestPath.length() > redirectPath.length()) {
+                return requestPath.substring(redirectPath.length());
+            }
+        }
+        return null;
+    }
+
+    private String generateState(RoutingContext context, String dynamicPath) {
+        String uuid = UUID.randomUUID().toString();
+        String cookieValue = uuid + COOKIE_DELIM + dynamicPath;
```

Comment:
You could probably reuse the state parameter to bind the callback uri.

Reply:
@pedroigor yeah, I thought about it but if it goes in the URL it be encoded so it may end up being double encoded and as such we can lose the original percent encoding if any in the original path, hence I added it to the cookie. Going forward I guess we can expect the cookie become a (secured) container of various properties...